### DISCUSSION TITLE
老大可以分享一下arm平台的交叉编译工具链和相关的Makefile这些配置吗？

### DISCUSSION THREAD

**Discussion by faajaudai** [pymumu/smartdns#1052] [General]
是这样，我有一个光猫用的是中兴(ZXIC)定制的ARM芯片，我一直想自己编译一些小玩意丢上去玩玩
但我在网上找了好几个预编译的ARM交叉编译工具链，编译个HelloWorld丢上去都跑不起来
刚刚下午发现这个项目里release的那些静态编译的arm平台程序，可以直接在光猫里跑起来
所以想问问大佬是否可以分享一下这个项目在交叉编译时的那个ARM的工具链和编译时的那些配置文件
我想学习一下交叉编译相关的技巧

**Comment by pymumu** [comment]
如果系统资源足够的话，推荐你用go，那个交叉编译很简单。

我的交叉编译工具是用的crosstool-ng生成的。你选择好对应的libc库，版本，就可以生成工具链了。
我的是树莓派上的工具链，估计你用不了。

**Reply by faajaudai** [reply]
好的，看来还是要从头啃，没有捷径，我研究一下crosstool-ng怎么玩。
你说你的工具链是树莓派上的，意思是你用crosstool-ng生成了一套在树莓派上跑的工具链，再在树莓派上做的本地编译吗？

Go的话好像是简单很多，但我看在路由器这种嵌入式设备上跑的项目，很多还是用C写的，碰到想用的工具时还是绕不开
我是学Java出身，对这块其实比较小白，看那些复杂一点的Makefile，里面的GCC那些选项、参数什么的，感觉都被绕晕TT

**Reply by pymumu** [reply]
smartdns是使用vscode，远程连接到树莓派上开发的。
我自己主要是将smartdns运行在华硕路由器上，因为路由器资源有限，才选用c开发的。
非嵌入式开发，或者是临时玩玩，建议直接用go，不用折腾gcc，makefile。交叉编译等。

如果实在是要类似嵌入式开发，可以用用rust，那个属于可以有机会替换C的现代语言。交叉编译也很简单。就是可能入门门槛高点。入门后，就没难度了。

我的交叉编译工具是在树莓派上用crosstool生成的，配置好后，编译就好了，树莓派比较慢，大概100分钟编译好。
x86的话，20多分钟就应该好了。这个只是配置编译，和编译openwrt固件类似，难度不高。

**Comment by PikuZheng** [comment]
有捷径啊！利用qemu和docker buildx，直接构建一个对应架构的环境，然后正常编译就可以啦！
